### PR TITLE
Fix curiosities in `TimerDescriptor`

### PR DESCRIPTION
**Pull request by alexey-milovidov**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

A lot of suspicious errors since 2022: https://github.com/ClickHouse/ClickHouse/issues/37686
And I, indeed, found a lot of potential bugs in the code. But they are most likely not related.

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

File: src/Common/TimerDescriptor.cpp [hunk 1]
```diff
@@ -19,15 +20,13 @@ namespace ErrorCodes
 
 TimerDescriptor::TimerDescriptor(int clockid, int flags)
 {
-    timer_fd = timerfd_create(clockid, flags);
+    timer_fd = timerfd_create(clockid, flags | TFD_NONBLOCK);
     if (timer_fd == -1)
-        throw Exception(ErrorCodes::CANNOT_CREATE_TIMER, "Cannot create timer_fd descriptor");
-
-    if (-1 == fcntl(timer_fd, F_SETFL, O_NONBLOCK))
-        throw ErrnoException(ErrorCodes::CANNOT_FCNTL, "Cannot set O_NONBLOCK for timer_fd");
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:27]: If exception happens here, the opened file description is leaked.

File: src/Common/TimerDescriptor.cpp [hunk 2]
```diff
@@ -19,15 +20,13 @@ namespace ErrorCodes
 
 TimerDescriptor::TimerDescriptor(int clockid, int flags)
 {
-    timer_fd = timerfd_create(clockid, flags);
+    timer_fd = timerfd_create(clockid, flags | TFD_NONBLOCK);
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:23]: It's possible to set O_NONBLOCK immediately.

File: src/Common/TimerDescriptor.cpp [hunk 3]
```diff
@@ -43,13 +42,16 @@ TimerDescriptor::~TimerDescriptor()
     /// Do not check for result cause cannot throw exception.
     if (timer_fd != -1)
     {
-        int err = close(timer_fd);
-        chassert(!err || errno == EINTR);
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:47]: `chassert` does not work in release builds.

File: src/Common/TimerDescriptor.cpp [hunk 4]
```diff
@@ -43,13 +42,16 @@ TimerDescriptor::~TimerDescriptor()
...
+        if (0 != ::close(timer_fd))
+            std::terminate();
     }
 }
 
 void TimerDescriptor::reset() const
 {
+    if (timer_fd == -1)
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:52]: Here, we didn't check the case of moved-out objects.

File: src/Common/TimerDescriptor.cpp [hunk 5]
```diff
@@ -66,25 +68,44 @@ void TimerDescriptor::reset() const
 
 void TimerDescriptor::drain() const
 {
+    if (timer_fd == -1)
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:71]: Here, we didn't check the case of moved-out objects.

File: src/Common/TimerDescriptor.cpp [hunk 6]
```diff
@@ -66,25 +68,44 @@ void TimerDescriptor::reset() const
 
 void TimerDescriptor::drain() const
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:69]: The code was difficult to read, I added comments.

File: src/Common/TimerDescriptor.cpp [hunk 7]
```diff
@@ -66,25 +68,44 @@ void TimerDescriptor::reset() const
...
+                continue;
+
+            throw ErrnoException(ErrorCodes::CANNOT_READ_FROM_SOCKET, "Cannot drain timer_fd");
         }
     }
 }
 
 void TimerDescriptor::setRelative(uint64_t usec) const
 {
+    chassert(timer_fd >= 0);
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:107]: This could help.

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 5ac26d2115c9e3e777990fc932bf03e83a2effee with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/67287/5ac26d2115c9e3e777990fc932bf03e83a2effee/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

File: src/Common/TimerDescriptor.cpp [hunk 8]
```diff
@@ -19,15 +20,13 @@ namespace ErrorCodes
 
 TimerDescriptor::TimerDescriptor(int clockid, int flags)
 {
-    timer_fd = timerfd_create(clockid, flags);
+    timer_fd = timerfd_create(clockid, flags | TFD_NONBLOCK);
     if (timer_fd == -1)
-        throw Exception(ErrorCodes::CANNOT_CREATE_TIMER, "Cannot create timer_fd descriptor");
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:24]: It's strange that here it was Exception, but next two lines, it was ErrnoException.

File: src/Common/TimerDescriptor.cpp [hunk 9]
```diff
@@ -17,17 +18,15 @@ namespace ErrorCodes
     extern const int CANNOT_READ_FROM_SOCKET;
 }
 
-TimerDescriptor::TimerDescriptor(int clockid, int flags)
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:20]: Removed useless arguments.

File: src/Common/TimerDescriptor.cpp [hunk 10]
```diff
@@ -17,17 +18,15 @@ namespace ErrorCodes
...
+TimerDescriptor::TimerDescriptor()
 {
-    timer_fd = timerfd_create(clockid, flags);
+    timer_fd = timerfd_create(CLOCK_MONOTONIC, TFD_NONBLOCK | TFD_CLOEXEC);
```
Comment (alexey-milovidov) [inline src/Common/TimerDescriptor.cpp:23]: Added CLOEXEC, it could be needed for UDFs.

**Pull request merged**